File: src/info/macos/cpu.rs
use std::{ffi::c_int, mem::size_of, ptr::null};

use libc::{c_void, difftime, sysctl, time, time_t, timeval, CTL_KERN, KERN_BOOTTIME};

use super::*;

pub fn cpu_info() -> Result<Cpu, InfoError> {
    let uptime = uptime()?;
    let name = cpu_name()?;
    let clock_rate = cpu_frequency()?;
    let (cores, threads) = cpu_count()?;

    Ok(Cpu {
        uptime,
        name,
        clock_rate,
        cores,
        threads,
    })
}

fn uptime() -> Result<u128, InfoError> {
    // Initialize the structure
    let mut boottime: timeval = timeval {
        tv_sec: 0,
        tv_usec: 0,
    };

    // Get hw.boottime
    let mut mib: [c_int; 2] = [CTL_KERN, KERN_BOOTTIME];
    if unsafe {
        sysctl(
            &mut mib as *mut c_int,
            2,
            &mut boottime as *mut timeval as *mut c_void,
            &mut size_of::<timeval>(),
            null::<usize>() as *mut c_void,
            0,
        )
    } < 0
    {
        return Err(InfoError::Sysctl {
            name: "hw.boottime".to_string(),
        });
    }
    
    // // Get the time elapsed (assuming the current time hasn't been changed)
    let bsec: time_t = boottime.tv_sec;
    let csec: time_t = unsafe { time(null::<i64>() as *mut i64) };
    let uptime = (unsafe { difftime(csec, bsec) } * 1000.0) as u128;

    Ok(uptime)
}

fn cpu_frequency() -> Result<f64, InfoError> {
    let mut frequency: u64 = 0;
    if unsafe { macos_cpu_frequency(&mut frequency as *mut u64) } < 0 {
        return Err(InfoError::General("".to_string()));
    }

    Ok(frequency as f64)
}

fn cpu_name() -> Result<String, InfoError> {
    let mut buffer = [0x0; 2048];

    if unsafe { macos_cpu_name(buffer.as_mut_ptr(), buffer.len()) } < 0 {
        return Err(InfoError::General("Unable to get CPU name".to_string()));
    }

    let buffer = buffer.map(|char| char as u8);

    let s = std::str::from_utf8(&buffer).expect("INVALID UTF-8");

    Ok(s.trim().to_string())
}

fn cpu_count() -> Result<(usize, usize), InfoError> {
    let mut buffer = MacOsCpuCount {
        core_count: 0,
        thread_count: 0,
    };

    if unsafe { macos_cpu_count(&mut buffer) } < 0 {
        return Err(InfoError::General("Unable to get CPU name".to_string()));
    }

    Ok((buffer.core_count, buffer.thread_count))
}

File: lib/info/macos/cpu.c
#include "cpu.h"

#ifdef __APPLE__
#include <string.h>
#include <sys/sysctl.h>
#include <sys/types.h>

int macos_cpu_name(char *buffer, size_t buffer_len)
{
    char brand_string[2048];
    size_t brand_string_size = sizeof(brand_string);
    if (sysctlbyname("machdep.cpu.brand_string", &brand_string, &brand_string_size, NULL, 0) < 0)
        return -1;

    if (buffer_len < strlen(brand_string) + 1)
        return -2;
    else
        strcpy(buffer, brand_string);
    return 0;
}

int macos_cpu_frequency(uint64_t* cpu_frequency)
{
    uint64_t frequency = 0;
    size_t frequency_size = sizeof(frequency);

    if (sysctlbyname("hw.cpufrequency", &frequency, &frequency_size, NULL, 0) < 0)
    {
        return -1;
    }
    *cpu_frequency = frequency / 1000000;
    return 0;
}

int macos_cpu_count(struct MacOsCpuCount* info)
{
    size_t core_count = 0, thread_count =0;
    size_t core_count_size = sizeof(core_count), thread_count_size = sizeof(thread_count);

    if (sysctlbyname("machdep.cpu.core_count", &core_count, &core_count_size, NULL, 0) < 0)
    {
        return -1;
    }

    if (sysctlbyname("machdep.cpu.thread_count", &thread_count, &thread_count_size, NULL, 0) < 0)
    {
        return -1;
    }

    info->core_count = core_count;
    info->thread_count = thread_count;
    return 0;
}

#else
int macos_cpu_frequency(uint64_t* cpu_frequency)
{
    return -1;
}

int macos_cpu_count(struct MacOsCpuCount* info)
{
    return -1;
}

int macos_cpu_name(char *buffer, size_t buffer_len)
{
    return -1;
}
#endif